extensions/amp-form/0.1/test-e2e/test-ssr.js

```diff
@@ -0,0 +1,50 @@
+/*e
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Amp-form test for standalone and in an SSR enabled viewer. Verifies that
+ * the rendered output is the same if the response from the REST endpoint is
+ * the same as from the SSR endpoint.
+ */
+
+describes.endtoend.only('AMP form server side rendered templates', {
+  testUrl: 'http://localhost:8000/test/fixtures/e2e/amp-form/amp-form.ssr.html',
+  environments: ['single', 'viewer-demo'],
+  experiments: ['layers'],
+}, async env => {
+  let controller;
+
+  beforeEach(async() => {
+    controller = env.controller;
+  });
+
+  it('should render server side render form response', async() => {
+    const searchInput =
+        await controller.findElement('#xhr-get  input[name=term]');
+    await controller.type(searchInput, 'search term');
+    const submitForm =
+        await controller.findElement('#xhr-get input[type=submit]');
+    await controller.click(submitForm);
+
+    const renderedTemplate = await controller.findElement(
+        'div[i-amphtml-rendered]');
+    await expect(controller.getElementText(renderedTemplate)).to.equal(
+        'Here are the results for the search:\nResult 1\nResult 2\nResult 3');
+
+    await controller.takeScreenshot('screenshots/amp-form-ssr.png');
+  });
+
```

Comment:
Can we add a test for the non-SSR case here as discussed?

Reply:
The tests already covers non ssr as the existing test runs for the environments single (non-ssr) and viewer-demo (ssr). Perhaps the image name needs to be renamed.

Reply:
Ah, they should be separate with different rendered output. Otherwise, how would we detect if SSR is accidentally bypassed?

Reply:
I think there is still some value to running the same test through the two different environments and expecting the same rendered output, as it tests that the two different rendering paths for the amp-components still yields the same output. Though yes, there should be additional tests added, testing only the specific environments and with different rendered output.

Reply:
Actually for the SSR case, I can add an expectation that the doc is in the viewer.

Reply:
talked to Will offline, we will address these test cases in the two separate environments separately.